### PR TITLE
Improve performance of readuntil using strings (v2)

### PR DESCRIPTION
**Pull request by vtjnash**
I was impressed by #20621, so once I understood the algorithm there, I decided to try improving it even further (imitation is the sincerest form of flattery?) by computing the backtracking lazily:

```julia
julia> str = String([rand('A':'Z', 50000); '1']);

julia> io = IOBuffer(str);

julia> goal = str[(end - 1000):end];

julia> function readuntil_old(s::IO, r::AbstractString)
                  l = length(r)
                  if l == 0
                      return ""
                  end
                  out = IOBuffer()
                  m = Array{Char}(l)  # last part of stream to match
                  t = collect(r)
                  i = 0
                  while !eof(s)
                      i += 1
                      c = read(s, Char)
                      write(out, c)
                      if i <= l
                          m[i] = c
                      else
                          # shift to last part of s
                          for j = 2:l
                              m[j-1] = m[j]
                          end
                          m[l] = c
                      end
                      if i >= l && m == t
                          break
                      end
                  end
                  return String(take!(out))
              end
readuntil_old (generic function with 1 method)

julia> b1 = @benchmark readuntil_old(seekstart($io), $goal);

julia> b2 = @benchmark Base.readuntil(seekstart($io), $goal);

julia> judge(median(b2), median(b1))
BenchmarkTools.TrialJudgement: 
  time:   -99.05% => improvement (5.00% tolerance)
  memory: -10.81% => improvement (1.00% tolerance)

julia> goal = str[end:end];

julia> b1 = @benchmark readuntil_old(seekstart($io), $goal);

julia> b2 = @benchmark Base.readuntil(seekstart($io), $goal);

julia> judge(median(b2), median(b1))
BenchmarkTools.TrialJudgement: 
  time:   -93.04% => improvement (5.00% tolerance)
  memory: +113.22% => regression (1.00% tolerance)

julia> goal = str[(end - 1):end];

julia> b1 = @benchmark readuntil_old(seekstart($io), $goal);

julia> b2 = @benchmark Base.readuntil(seekstart($io), $goal);

julia> judge(median(b2), median(b1))
BenchmarkTools.TrialJudgement: 
  time:   -71.47% => improvement (5.00% tolerance)
  memory: -0.21% => invariant (1.00% tolerance)

julia> goal = str;

julia> b1 = @benchmark readuntil_old(seekstart($io), $goal);

julia> b2 = @benchmark Base.readuntil(seekstart($io), $goal);

julia> judge(median(b2), median(b1))
BenchmarkTools.TrialJudgement: 
  time:   -82.88% => improvement (5.00% tolerance)
  memory: -85.44% => improvement (1.00% tolerance)
```

Note that `Base.readuntil` is the algorithm in this PR, and the one memory regression reported is a measurement / reporting inaccuracy due to the algorithm used by `readuntil(io::AbstractIOBuffer, delim::UInt8)` (afaict).

### CODE REVIEW

File: base/io.jl [hunk 1]
```diff
@@ -115,8 +115,6 @@ unsafe_read(io::AbstractPipe, p::Ptr{UInt8}, nb::UInt) = unsafe_read(pipe_reader
 read(io::AbstractPipe) = read(pipe_reader(io))
 readuntil(io::AbstractPipe, arg::UInt8)          = readuntil(pipe_reader(io), arg)
 readuntil(io::AbstractPipe, arg::Char)           = readuntil(pipe_reader(io), arg)
-readuntil(io::AbstractPipe, arg::AbstractString) = readuntil(pipe_reader(io), arg)
-readuntil(io::AbstractPipe, arg)                 = readuntil(pipe_reader(io), arg)
```
Comment (tkelman) [inline base/io.jl]: what's the replacement for this?

Comment (vtjnash) [inline base/io.jl]: none needed. these methods just cause ambiguity conflicts, they usually can't be called.

**Review by omus — COMMENTED** [review]
Nice. I'll want to compare the performance of this to what I originally wrote. One thing I was concerned about in my original code was recomputing the position when we need to backtrack several times.

File: base/io.jl [hunk 2]
```diff
@@ -445,39 +443,80 @@ function readuntil{T}(s::IO, delim::T)
...
-    l = length(t)
-    if l == 0
+function _rfind_sub_c(target, endc, endstate)
+    # find the largest i (< endstate) such that "target[(end - i + 1):(end - 1)] * c" == "target[1:i]"
+    # (equivalently, find the smallest tii' such that "target[tii':(end - 1)] * c" == "target[1:(end - tii' + 1)]")
+    # this assumes that the `target` iterator is pure and supports equality comparison for its state
+    # and that endstate is a valid state for target
+    tii = start(target)
+    i = start(target)
+    ti = tii
```
Comment (vtjnash) [inline base/io.jl]: I wrote it this way because it mirrors the ordering needed to update the iterator in the while loop

File: base/io.jl [hunk 3]
```diff
@@ -445,39 +443,80 @@ function readuntil{T}(s::IO, delim::T)
...
+    while tii != endstate
```
Comment (omus) [inline base/io.jl]: Interesting. Out of curiosity are there any iterators that aren't ordered?

**Comment by vtjnash** [comment]
> Nice. I'll want to compare the performance of this to what I originally wrote. One thing I was concerned about in my original code was recomputing the position when we need to backtrack several times.

Yes, the worse-case here is worse. I think the average-case should be better though since I think backtracking will usually be fairly rare and it'll usually be quick to compute the specific backtrack value on-demand.

Looking back at the PR that implemented `readuntil`, we can see that your algorithm is KMP (https://github.com/JuliaLang/julia/issues/1684#issuecomment-11262134) :). I think my version is `O(n*k)` (worst case), `O(n)` (best case), `O(n + k)` (average case); while yours is `O(n + k)` (constant).

Where I think the worse case is of the form `target = "c"^k * "d"` with `string = "c"^n`, where it must do `O(k)` work computing the backtracking for each of the `O(n)` elements in the string. For the average case, I am assuming the the number of backtracking attempts is independent of the length of the string (i.e. that `P(match(n) | match(n - 1)) >> P(match(n))`, i.e. that the strings aren't i.i.d.)

File: base/io.jl (hunk 2 above)
Comment (StefanKarpinski) [inline base/io.jl]: Even though `start` tends to be cheap, there's no point in calling it twice.

**Comment by StefanKarpinski** [comment]
Is there any way this version could degrade gracefully to the `O(n + k)` version?

**Comment by omus** [comment]
I was thinking about this and maybe there is a way to get the worst case down to `O(n + k)`. My version computed and stored the backtracking information upfront while this version computes the backtracking info on-the-fly.

If instead we cache the on-the-fly computations we should be able to improve the worst case with a small memory cost.

**Comment by omus** [comment]
The work here was integrated into #20621

**Pull request closed without merging**